packages/graphql-language-service-server/src/languages.ts

```diff
@@ -0,0 +1,66 @@
+import { Position, Range } from 'graphql-language-service-utils';
+import { makeExtractTagsFromSource, GraphQLSource } from './languageUtils';
+
+export type LanguageConfig = {
+  readonly extensions: ReadonlyArray<string>;
+  readonly fileFilter: (content: string) => boolean;
+  readonly extractSources: (content: string) => ReadonlyArray<GraphQLSource>;
+};
+
+/**
+ * Filters and other regexps used later on.
+ */
+const jsFileFilter = new RegExp(/graphql`|graphql\.experimental`|gql`/);
+const jsGraphQLTagsRegexp = new RegExp(
+  /(?<=(graphql|gql|graphql\.experimental)`)[\s\S]+?(?=`)/g,
+);
+
+// This removes interpolation in JS strings, like ` ${someInterpolationHere}`
+const jsRemoveInterpolationRegexp = new RegExp(/([\s]+\${)[\s\S]+?(})/g);
+
+const jsExtractor = makeExtractTagsFromSource(jsGraphQLTagsRegexp);
+
+const reasonFileFilterRegexp = new RegExp(/(\[%(graphql|relay\.))/);
+const reasonGraphQLTagsRegexp = new RegExp(
+  /(?<=\[%(graphql|relay\.\w*)[\s\S]*{\|)[.\s\S]+?(?=\|})/g,
+);
+const reasonExtractor = makeExtractTagsFromSource(reasonGraphQLTagsRegexp);
+
+const jsLanguages: LanguageConfig = {
+  extensions: ['js', 'jsx', 'ts', 'tsx'],
+  fileFilter: (content: string) => jsFileFilter.test(content),
+  extractSources: (content: string) =>
+    jsExtractor(content).map(source => ({
+      ...source,
+      template: source.template.replace(jsRemoveInterpolationRegexp, ''),
+    })),
+};
+
+const graphql: LanguageConfig = {
+  extensions: ['graphql'],
+  fileFilter: () => true,
+  extractSources: (content: string) => {
+    const lines = content.split('\n');
+    const range = new Range(
+      new Position(0, 0),
+      new Position(lines.length - 1, lines[lines.length - 1].length - 1),
+    );
+    return [{ template: content, range }];
+  },
+};
+
+const reasonML: LanguageConfig = {
+  extensions: ['re'],
+  fileFilter: (content: string) => reasonFileFilterRegexp.test(content),
+  extractSources: reasonExtractor,
+};
+
+export const languages: ReadonlyArray<LanguageConfig> = [
+  jsLanguages,
```

Comment:
so instead of readonly, why not export a function that pushes to this const array in memory, so someone could register some languages by calling this globally even, when instantiating the server seperately from the server instantiating config?
```ts
function registerLanguage(lang: LanguageConfig) {
     languages.push(lang)
} 
```
then, whenever `getLanguageFromExtension` is invoked the new languages are there

Reply:
Great idea! That'll make it very flexible. Adding it! :)